DependencyInjection/Compiler/EntityListenerPass.php

```diff
@@ -39,20 +39,50 @@ public function process(ContainerBuilder $container)
                     continue;
                 }
 
-                $resolver = sprintf('doctrine.orm.%s_entity_listener_resolver', $name);
-                if ($container->hasAlias($resolver)) {
-                    $resolver = (string) $container->getAlias($resolver);
+                $resolverId = sprintf('doctrine.orm.%s_entity_listener_resolver', $name);
+                if ($container->hasAlias($resolverId)) {
+                    $resolverId = (string) $container->getAlias($resolverId);
                 }
 
-                if (!$container->hasDefinition($resolver)) {
+                if (!$container->hasDefinition($resolverId)) {
                     continue;
                 }
 
+                $resolver = $container->getDefinition($resolverId);
+
                 if (isset($attributes['entity']) && isset($attributes['event'])) {
                     $this->attachToListener($container, $name, $id, $attributes);
                 }
 
-                $container->getDefinition($resolver)->addMethodCall('register', array(new Reference($id)));
+                if (isset($attributes['lazy']) && $attributes['lazy']) {
+                    $listener = $container->findDefinition($id);
+
+                    if (!$listener->isPublic()) {
+                        throw new \InvalidArgumentException(sprintf('The service "%s" must be public as this entity listener is lazy-loaded.', $id));
```

Comment:
I suggest using the InvalidArgumentException of the DI component instead, to be consistent with Symfony